src/D2L.CodeStyle.Analyzers/ApiUsage/Logging/LoggingExecutionContextScopeBuilderAnalyzer.cs

```diff
@@ -0,0 +1,147 @@
+﻿using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Linq;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.CSharp;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using Microsoft.CodeAnalysis.Diagnostics;
+
+namespace D2L.CodeStyle.Analyzers.ApiUsage.Logging {
+
+	[DiagnosticAnalyzer( LanguageNames.CSharp )]
+	internal sealed class LoggingExecutionContextScopeBuilderAnalyzer : DiagnosticAnalyzer {
+
+		public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics => ImmutableArray.Create(
+			Diagnostics.LoggingContextRunAwaitable
+		);
+
+		public override void Initialize( AnalysisContext context ) {
+			context.EnableConcurrentExecution();
+			context.RegisterCompilationStartAction( RegisterScopeBuilderAnalyzer );
+		}
+
+		private void RegisterScopeBuilderAnalyzer( CompilationStartAnalysisContext context ) {
+			var ILoggingExecutionContextScopeBuilder = context.Compilation.GetTypeByMetadataName( "D2L.LP.Logging.ExecutionContexts.ILoggingExecutionContextScopeBuilder" );
+
+			// ILoggingExecutionContextScopeBuilder not in compilation, no need to register
+			if( ILoggingExecutionContextScopeBuilder == null || ILoggingExecutionContextScopeBuilder.Kind == SymbolKind.ErrorType ) {
+				return;
+			}
+
+			ImmutableHashSet<IMethodSymbol> ILoggingExecutionContextScopeBuilderRunSymbols = ILoggingExecutionContextScopeBuilder
+				.GetMembers()
+				.Where( m => m.Kind == SymbolKind.Method )
+				.Where( m => m.MetadataName == "Run" )
+				.Cast<IMethodSymbol>()
+				.ToImmutableHashSet();
+
+			// Can't find ILoggingExecutionContextScopeBuilder.Run or ILoggingExecutionContextScopeBuilder.Run<T>
+			if( !ILoggingExecutionContextScopeBuilderRunSymbols.Any() ) {
+				return;
+			}
+
+			ImmutableHashSet<INamedTypeSymbol> taskTypeBuiltins = ImmutableHashSet
+				.Create(
+					context.Compilation.GetTypeByMetadataName( "System.Threading.Tasks.Task" ),
+					context.Compilation.GetTypeByMetadataName( "System.Threading.Tasks.Task`1" ),
+					context.Compilation.GetTypeByMetadataName( "System.Threading.Tasks.ValueTask" ),
+					context.Compilation.GetTypeByMetadataName( "System.Threading.Tasks.ValueTask`1" )
+				)
+				.Where( x => x != null && x.Kind != SymbolKind.ErrorType )
+				.ToImmutableHashSet();
+
+			INamedTypeSymbol AsyncMethodBuilderAttribute = context.Compilation.GetTypeByMetadataName( "System.Runtime.CompilerServices.AsyncMethodBuilderAttribute" );

```

Comment:
This might not be the right approach, since it might not be directly referenced by the assembly, but annotating a return type of another assembly? Don't know how that works.

Reply:
Consider adding a comment explaining what this is for; `[AsyncMethodBuilder]` is pretty obscure